File: q101-125/q113.py
# 2020.01.16
# Problem Statement:
# https://leetcode.com/problems/path-sum-ii/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root: return []
        self.sum = sum
        self.answer_list = []
        self.pathSumHelper(root, 0, [])
        return self.answer_list
    
    def pathSumHelper(self, root, sum, one_answer):        
        if not root: return
        sum = sum + root.val
        if sum == self.sum and self.isLeaf(root):
            self.answer_list.append(one_answer+[root.val])
            return
        else:
            # one_answer has to be passed in with operations inside function call, otherwise the value could be changed
            self.pathSumHelper(root.left, sum, one_answer+[root.val])
            self.pathSumHelper(root.right, sum, one_answer+[root.val])
    
    def isLeaf(self, root):
        if not root.left and not root.right: return True
        return False

File: Amazon/q113.java
// 2022.06.02
// Problem Statement:
// https://leetcode.com/problems/path-sum-ii/

// idea: recursion, search left and/or right, with targetSum-root.val
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        // base case 1
        if (root==null) return new ArrayList<> ();
        List<List<Integer>> answer = new ArrayList<> ();
        // base case 2
        if (root.left==null && root.right==null) {
            if (targetSum==root.val) {
                List<Integer> one_node = new ArrayList<> ();
                one_node.add(root.val);
                answer.add(one_node);
                return answer;
            } else {
                return answer;
            }
        }
        if (root.left!=null) {
            List<List<Integer>> left = pathSum(root.left, targetSum-root.val);
            for (List l_list : left) {
                l_list.add(0, root.val);
                answer.add(l_list);
            }
        }
        if (root.right!=null) {
            List<List<Integer>> right = pathSum(root.right, targetSum-root.val);
            for (List r_list : right) {
                r_list.add(0, root.val);
                answer.add(r_list);
            }
        }
        return answer;
    }
}